File: task/src/main/scala/com/goexp/galgame/data/piplline/core/MessageQueueProxy.java
package com.goexp.galgame.data.piplline.core;

import com.goexp.galgame.data.piplline.exception.RuntimeInterruptedException;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

public class MessageQueueProxy<T> {
    final private BlockingQueue<T> msgQueue;

    public MessageQueueProxy(int capacity) {
        msgQueue = new ArrayBlockingQueue<>(capacity);
    }

    public boolean offer(T o) {
        try {
            return msgQueue.offer(o, 60, TimeUnit.MINUTES);
        } catch (InterruptedException e) {
            e.printStackTrace();
            throw new RuntimeInterruptedException(e);
        }
    }

    public boolean offer(T o, long timeout, TimeUnit unit) {
        try {
            return msgQueue.offer(o, timeout, unit);
        } catch (InterruptedException e) {
            e.printStackTrace();
            throw new RuntimeInterruptedException(e);
        }
    }


    public T pull() {
        try {
            return msgQueue.poll(60, TimeUnit.MINUTES);
        } catch (InterruptedException e) {
            e.printStackTrace();
            throw new RuntimeInterruptedException(e);
        }
    }

    public T poll(long timeout, TimeUnit unit) {
        try {
            return msgQueue.poll(timeout, unit);
        } catch (InterruptedException e) {
            e.printStackTrace();
            throw new RuntimeInterruptedException(e);
        }
    }
}

File: task/src/main/scala/com/goexp/galgame/data/piplline/core/MessageQueueProxy.scala
package com.goexp.galgame.data.piplline.core

import java.util.concurrent.{ArrayBlockingQueue, TimeUnit}

import com.goexp.galgame.data.piplline.exception.RuntimeInterruptedException

class MessageQueueProxy[T <: Message](val capacity: Int) {
  private val msgQueue = new ArrayBlockingQueue[T](capacity)

  def offer(o: T): Boolean = try msgQueue.offer(o, 60, TimeUnit.MINUTES)
  catch {
    case e: InterruptedException =>
      e.printStackTrace()
      throw new RuntimeInterruptedException(e)
  }

  def offer(o: T, timeout: Long, unit: TimeUnit): Boolean = try msgQueue.offer(o, timeout, unit)
  catch {
    case e: InterruptedException =>
      e.printStackTrace()
      throw new RuntimeInterruptedException(e)
  }

  def pull: T = try msgQueue.poll(60, TimeUnit.MINUTES)
  catch {
    case e: InterruptedException =>
      e.printStackTrace()
      throw new RuntimeInterruptedException(e)
  }

  def poll(timeout: Long, unit: TimeUnit): T = try msgQueue.poll(timeout, unit)
  catch {
    case e: InterruptedException =>
      e.printStackTrace()
      throw new RuntimeInterruptedException(e)
  }
}